[WPE] Add new post-commit bots for performance tests with RPis

clopez opened the pull request:
#### e7e596ed660c38a7dbd1940cdd1ccf8aa34b08de
<pre>
[WPE] Add new post-commit bots for performance tests with RPis
<a href="https://bugs.webkit.org/show_bug.cgi?id=256283">https://bugs.webkit.org/show_bug.cgi?id=256283</a>

Reviewed by Jonathan Bedard.

This adds new bots for running performance tests on WPE with RPi4 boards.
The configuration is the following:

1) Build-only bots (x86_64 machines that cross-build)
 - WPE-Linux-RPi4-32bits-Mesa-Release-Perf-Build
   (builds with --cross-target=rpi4-32bits-mesa)
 - WPE-Linux-RPi4-64bits-Mesa-Release-Perf-Build
   (builds with --cross-target=rpi4-64bits-mesa)

2) Test-only bots (RPi4 boards that run the tests)
 - WPE-Linux-RPi4-32bits-Mesa-Release-Perf-Tests
 - WPE-Linux-RPi4-64bits-Mesa-Release-Perf-Tests

There will be 8 workers (RPi4 boards) connected per
tester (so a total of 16 RPi4 boards).

The bots use the standard BuildOnly and DownloadAndPerfTest factories
with a few extra steps for handling the cross-target setup.

And for handling everything related with the cross builds and the generation
and deployment of target images it uses the cross-toolchain-helper script
and related functionality that was introduced at 259479@main

The append*Flags() functions at steps.py are moved into a common MixIn class

* Tools/CISupport/build-webkit-org/config.json:
* Tools/CISupport/build-webkit-org/factories.py:
(Factory):
(Factory.__init__):
(BuildFactory.__init__):
(NoInstallDependenciesBuildFactory):
(CrossTargetBuildFactory):
(DownloadAndPerfTestFactory.__init__):
(CrossTargetDownloadAndPerfTestFactory):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/build-webkit-org/steps.py:
(appendCrossTargetFlags):
(appendCustomBuildFlags):
(appendCustomTestingFlags):
(InstallWpeDependencies.start):
(ArchiveBuiltProduct):
(ArchiveBuiltProduct.start):
(ShowIdentifier.hideStepIf):
(CheckIfNeededUpdateDeployedCrossTargetImage):
(CheckIfNeededUpdateDeployedCrossTargetImage.start):
(CheckIfNeededUpdateDeployedCrossTargetImage.evaluateCommand):
(CheckIfNeededUpdateRunningCrossTargetImage):
(CheckIfNeededUpdateRunningCrossTargetImage.evaluateCommand):
(BuildAndDeployCrossTargetImage):
(BuildAndDeployCrossTargetImage.start):
(RebootWithUpdatedCrossTargetImage):
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/264798@main">https://commits.webkit.org/264798@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/881c26e62caa7b101b90ed5efb352de07643e1b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10204 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8575 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8557 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8803 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11442 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8694 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9725 "3 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7684 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10359 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7025 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7815 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15357 "2 flakes 116 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11311 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8443 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6898 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/8499 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7720 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11929 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1021 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8185 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -49,6 +49,36 @@
 Interpolate = properties.Interpolate
 
 
+def appendCrossTargetFlags(step, additionalArguments):
```
We usually try to scope functions like this into a `MixIn` class.

clopez replied:
AFAIK the current code on `Tools/CISupport/build-webkit-org` does not use (still) mixin classes. But the code for EWS at `Tools/CISupport/ews-build` does

There are 3 more similar functions to this one without a class there: `appendCustomBuildFlags`, `appendCustomTestingFlags`, and `appendCrossTargetFlags`. This patch moves all the 3 of them to the top of the file so they can be referenced from any class below.

Do you think I should create something like a `CustomFlags` mixing class and move the 3 functions inside that class and then inherit from that class when needed? I can do it, just checking if that is what you suggest here

JonWBedard commented on Tools/CISupport/build-webkit-org/factories.py:
```diff
@@ -27,6 +27,9 @@
 
 class Factory(factory.BuildFactory):
+    ShouldInstallDependencies = True
+    ShouldUseCrossTargetImage = False
```
Style nit: Looks like we don't capitalize class variables like this, so this should be:
```
shouldInstallDependencies = True
shouldUseCrossTargetImage = False
```

clopez replied:
Ok, will change it

clopez replied:
It turns out in this file there were already previous variables starting with uppercase `Should` .. will change them also for consistency

clopez committed:
[WPE] Add new post-commit bots for performance tests with RPis
https://bugs.webkit.org/show_bug.cgi?id=256283

Reviewed by Jonathan Bedard.

This adds new bots for running performance tests on WPE with RPi4 boards.
The configuration is the following:

1) Build-only bots (x86_64 machines that cross-build)
 - WPE-Linux-RPi4-32bits-Mesa-Release-Perf-Build
   (builds with --cross-target=rpi4-32bits-mesa)
 - WPE-Linux-RPi4-64bits-Mesa-Release-Perf-Build
   (builds with --cross-target=rpi4-64bits-mesa)

2) Test-only bots (RPi4 boards that run the tests)
 - WPE-Linux-RPi4-32bits-Mesa-Release-Perf-Tests
 - WPE-Linux-RPi4-64bits-Mesa-Release-Perf-Tests

There will be 8 workers (RPi4 boards) connected per
tester (so a total of 16 RPi4 boards).

The bots use the standard BuildOnly and DownloadAndPerfTest factories
with a few extra steps for handling the cross-target setup.

And for handling everything related with the cross builds and the generation
and deployment of target images it uses the cross-toolchain-helper script
and related functionality that was introduced at 259479@main

The append*Flags() functions at steps.py are moved into a common MixIn class

* Tools/CISupport/build-webkit-org/config.json:
* Tools/CISupport/build-webkit-org/factories.py:
(Factory):
(Factory.__init__):
(BuildFactory.__init__):
(NoInstallDependenciesBuildFactory):
(CrossTargetBuildFactory):
(DownloadAndPerfTestFactory.__init__):
(CrossTargetDownloadAndPerfTestFactory):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/build-webkit-org/steps.py:
(appendCrossTargetFlags):
(appendCustomBuildFlags):
(appendCustomTestingFlags):
(InstallWpeDependencies.start):
(ArchiveBuiltProduct):
(ArchiveBuiltProduct.start):
(ShowIdentifier.hideStepIf):
(CheckIfNeededUpdateDeployedCrossTargetImage):
(CheckIfNeededUpdateDeployedCrossTargetImage.start):
(CheckIfNeededUpdateDeployedCrossTargetImage.evaluateCommand):
(CheckIfNeededUpdateRunningCrossTargetImage):
(CheckIfNeededUpdateRunningCrossTargetImage.evaluateCommand):
(BuildAndDeployCrossTargetImage):
(BuildAndDeployCrossTargetImage.start):
(RebootWithUpdatedCrossTargetImage):
* Tools/CISupport/build-webkit-org/steps_unittest.py:

Canonical link: https://commits.webkit.org/264798@main
